core/src/com/unciv/ui/mapeditor/TileEditorOptionsTable.kt

```diff
@@ -149,48 +153,92 @@ class TileEditorOptionsTable(val mapEditorScreen: MapEditorScreen): Table(Camera
             }
             improvementsTable.add(improvementImage).row()
         }
-        editorPickTable.add(AutoScrollPane(improvementsTable)).height(scrollPanelHeight)
+        editorPickTable.add(AutoScrollPane(improvementsTable).apply { setScrollingDisabled(true,false) }).height(scrollPanelHeight)
 
-        val nationsTable = Table()
-        for(nation in ruleset.nations.values){
-            val nationImage = getHex(Color.WHITE, ImageGetter.getNationIndicator(nation, 40f))
-            nationImage.onClick {
-                val improvementName = "StartingLocation "+nation.name
+        val nationTable = Table()
 
-                tileAction = {
-                    it.improvement = improvementName
-                    for (tileGroup in mapEditorScreen.mapHolder.tileGroups.values) {
-                        val tile = tileGroup.tileInfo
-                        if (tile.improvement == improvementName && tile != it)
-                            tile.improvement = null
-                        tile.setTerrainTransients()
-                        tileGroup.update()
+        if (UncivGame.Current.scenarioDebugSwitch) {
+            /** new scenario/improvements functionality
+             * There shoudn't be random players in scenario
+             * */
+            for (player in gameParameters.players) {
+                val playerIndex = gameParameters.players.indexOf(player) + 1
+                    if (player.chosenCiv == "Random") {
+                    } else {
+                        val nation = ruleset.nations[player.chosenCiv]!!
+                        val nationImage = ImageGetter.getNationIndicator(nation, 40f)
+                        nationImage.onClick {
+                            val improvementName = "StartingLocation " + nation.name
+
+                            tileAction = {
+                                it.improvement = improvementName
+                                for (tileGroup in mapEditorScreen.mapHolder.tileGroups.values) {
+                                    val tile = tileGroup.tileInfo
+                                    if (tile.improvement == improvementName && tile != it)
+                                        tile.improvement = null
+                                    tile.setTerrainTransients()
+                                    tileGroup.update()
+                                }
+                            }
```

Comment:
The tileAction seems the same for both cases, so you could define it outside of the if/else and have it only once

Reply:
there is `val improvementName` which is calculated for each nation. Maybe we can pass tileAction as a function of nation or nation.name?